### PR TITLE
feat(versions,nix): add 0_2 and necessary nix changes

### PR DESCRIPTION
**Pull request by steveeJ**
### TODO:
- [x] re-evaluate flake structure due the flake.lock issue ineffective locking of the versions
- [x] adapt derivations for 0.2 component versions
- [x] clean up sweep
- [x] exclude/fix failing tests
- [x] clean up sweep
- [x] remove sqlite dependency (requires bumping holochain in launcher and scaffolding)
    - [x] https://github.com/holochain/scaffolding/pull/94
- [x] dry-run
    - [x] https://github.com/holochain/holochain/actions/runs/4939253691
    - [ ] https://github.com/holochain/holochain/actions/runs/4945309693/jobs/8841936023 (failed due to a flaky test on x86_64-darwin)
- [ ] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

File: nix/modules/scripts.nix [hunk 1]
```diff
@@ -35,17 +35,86 @@
...
-        cd versions/0_1
-        nix flake update
-        cd ../../
+        export VERSIONS_DIR="./versions/''${1}"
+        export DEFAULT_VERSIONS_DIR="$(nix flake metadata --no-write-lock-file --json | jq --raw-output '.locks.nodes.versions.locked.path')"
 
-        nix flake lock --update-input versions --override-input versions "path:./versions/0_1"
+        (
+          cd "$VERSIONS_DIR"
+          nix flake update --tarball-ttl 0
```
Comment (steveeJ) [inline nix/modules/scripts.nix:43]: try out this for an atomic update procedure. it's unclear how this affects clients, as they don't have access to the path

```suggestion
          nix flake update --tarball-ttl 0 --override-input holochain-flake ../..
```

File: holonix/examples/custom_versions/flake.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{
+  # used for debugging
+
+  # useful commands:
+  # show which tag holochain points to:
+  # git tag --points-at $(nix flake metadata . --json | jq --raw-output '.locks.nodes.holochain.locked.rev') | grep holochain-
+
+  description = "Template for Holochain app development with custom versions";
+
+  inputs.holochain-flake.url = "github:holochain/holochain/pr_versions_0_2?dir=versions/0_2";
```
Comment (steveeJ) [inline holonix/examples/custom_versions/flake.nix:10]:
```suggestion
  inputs.holochain-flake.url = "github:holochain/holochain?dir=versions/0_2";
```

File: .github/workflows/release-prepare.yml [hunk 3]
```diff
@@ -201,32 +188,9 @@ jobs:
...
-                --log-level=debug \
-                release \
-                  --dry-run \
-                  --no-verify \
-                  --steps=PublishToCratesIo
-
-            cargo sweep -f
-            '
+          nix run .#cargo-sweep -s
+          nix run .#scripts-release-automation-prepare $PWD
```
Comment (steveeJ) [inline .github/workflows/release-prepare.yml:192]: double-check this after the rename

File: .github/workflows/release-prepare.yml [hunk 4]
```diff
@@ -169,26 +169,13 @@ jobs:
...
-              crate detect-missing-releaseheadings
-
       - name: Generate crate READMEs from doc comments
         if: ${{ inputs.skip_prepare_logic != 'true' }}
         env:
           HOLOCHAIN_REPO: ${{ inputs.HOLOCHAIN_REPO }}
         run: |
           cd "${HOLOCHAIN_REPO}"
-          nix develop .#release --ignore-environment --command bash ./scripts/generate_readmes.sh
+          nix run .#scripts-ci-generate-readmes
```
Comment (steveeJ) [inline .github/workflows/release-prepare.yml:178]: could also be moved to the check-and-prepare derivation

File: .github/workflows/holochain-build-and-test.yml [hunk 5]
```diff
@@ -34,10 +43,17 @@ jobs:
...
-              extra_arg: "--override-input holochain ${{ inputs.repo_path }}"
+                - build-release-automation-tests-unit
+              extra_arg: "--override-input versions 'path:./versions/0_2' --override-input holochain ${{ inputs.repo_path }}"
+            platform:
+              system: x86_64-linux
+
+          # FIXME: this can be skipped during a release run because it duplicates the release-prepare job
+          - cmd:
+              pkgs:
+                - build-release-automation-tests-repo
```
Comment (steveeJ) [inline .github/workflows/holochain-build-and-test.yml:55]: double-check name

File: nix/modules/launcher.nix [hunk 6]
```diff
@@ -19,41 +18,45 @@
 
         cargoExtraArgs = "--bin hc-launch";
 
-        buildInputs =
+        buildInputs = (with pkgs; [
+          openssl
+
+          # TODO: remove this once the features have been rearranged to use vendored sqlite
+          sqlite
```
Comment (steveeJ) [inline nix/modules/launcher.nix:25]: can be avoided by bumping holochain in the launcher to include these respectively:
* 0.1 https://github.com/holochain/holochain/commit/6de1fb217679bf0948769946652aebcea0df8876
* 0.2 https://github.com/holochain/holochain/commit/ca440c705eef2a2351a63d4da849ed4376c2222c

File: nix/modules/scaffolding.nix [hunk 7]
```diff
@@ -21,24 +20,26 @@
 
         buildInputs =
           (with pkgs; [
-            # TODO: remove sqlite package once https://github.com/holochain/holochain/pull/2248 is released
             openssl
+
+            # TODO: remove sqlite package once https://github.com/holochain/holochain/pull/2248 is released
             sqlite
```
Comment (steveeJ) [inline nix/modules/scaffolding.nix:26]: can be avoided by bumping holochain in the launcher to include these respectively:
* 0.1 https://github.com/holochain/holochain/commit/6de1fb217679bf0948769946652aebcea0df8876
* 0.2 https://github.com/holochain/holochain/commit/ca440c705eef2a2351a63d4da849ed4376c2222c

File: nix/modules/scripts.nix [hunk 8]
```diff
@@ -35,16 +35,34 @@
...
+          exit 0
+        else
+          git add "$VERSIONS_DIR"/flake.lock
         fi
+
+        if [[ "$VERSIONS_DIR" = "$DEFAULT_VERSIONS_DIR" ]]; then
+          nix flake lock --tarball-ttl 0 --update-input versions --override-input versions "path:$VERSIONS_DIR" 
+        fi
+
+        if [[ $(${pkgs.git}/bin/git diff -- flake.lock | grep -E '^[+-]\s+"' | grep -v lastModified --count) -eq 0 ]]; then
```
Comment (ThetaSinner) [inline nix/modules/scripts.nix:58]: Is there a difference between `${pkgs.git}/bin/git` and `git`?

**Comment by ThetaSinner** [comment]
I see you've already started removing sqlite, that would be my other comment - more recent builds of holochain/tools will not need it

File: nix/modules/scripts.nix (hunk 8 above)
Comment (steveeJ) [reply nix/modules/scripts.nix:58]: yes, just `git` is impure and relies on `git` being available in `PATH`. `${pkgs.git}/bin/git` evaluates to an absolute path to the binary.

Comment (steveeJ) [reply nix/modules/scripts.nix:58]: that said, i haven't been consistent in using one or the other.

**Review by ThetaSinner — APPROVED** [review]
Looking good!

**Pull request merged**